### ISSUE TITLE
Crash with wayland enabled on Gnome Wayland and display scaling enabled

### ISSUE THREAD

**Issue by szbergeron** [wez/wezterm#1727]
### What Operating System(s) are you seeing this problem on?

Linux Wayland

### WezTerm version

wezterm 20220317-165235-27088fab

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

``` 2022-03-18T01:41:36.784Z INFO  wezterm_mux_server_impl::local > setting up /run/user/1000/wezterm/gui-sock-2494986
Unsupported modifier, resource creation failed.
XXX: resource creation failed
 2022-03-18T01:41:36.947Z INFO  wezterm_gui::termwindow        > OpenGL initialized! Mesa Intel(R) UHD Graphics (CML GT2) 4.6 (Compatibility Profile) Mesa 22.0.0 is_context_loss_possible=false wezterm version: 20220317-165235-27088fab
Unsupported modifier, resource creation failed.
XXX: resource creation failed
 2022-03-18T01:41:37.042Z ERROR wezterm_gui                    > Error while flushing display: Broken pipe (os error 32); terminating
thread 'main' panicked at 'cannot access a Thread Local Storage value during or after destruction: AccessError', /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/local.rs:388:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fish: Job 1, 'cargo run --release --bin wezte…' terminated by signal SIGILL (Illegal instruction)```

### To Reproduce

1. Enable wayland
2. Attempt to launch the application

### Configuration

local wezterm = require 'wezterm';

return {
    enable_wayland = true,
}


### Expected Behavior

Application launches and displays a window

### Logs

_No response_

### Anything else?

_No response_

**Comment by wez** [comment]
Sounds like this error is happening inside Mesa:
https://github.com/mesa3d/mesa/blob/52ceb9dcb6f161295c28d37fa17dfe06d838fcaa/src/gallium/drivers/iris/iris_resource.c#L1067

wezterm isn't operating directly at the drm level, so this sounds like it might be a mesa-internal issue.

What can you tell me about the graphics hardware in your system?

Does `dmesg` have any logs relating to `drm` that might be relevant?

I think you should be able to run in software rendering mode as a workaround; try setting `front_end = "Software"` in your config file (https://wezfurlong.org/wezterm/config/lua/config/front_end.html).

**Comment by unrelentingtech** [comment]
hm. What does `weston-info` say about dmabuf formats on gnome vs sway?

**Comment by szbergeron** [comment]
@unrelentingtech 

wayland-info on gnome:
`interface: 'wl_compositor',                              version:  5, name:  1
interface: 'wl_drm',                                     version:  2, name:  2
interface: 'wl_shm',                                     version:  1, name:  3
	formats: 'XB4H'(0x48344258) 'AB4H'(0x48344241) 'XR4H'(0x48345258) 'AR4H'(0x48345241) 'XB30'(0x30334258) 'AB30'(0x30334241) 'XR30'(0x30335258) 'AR30'(0x30335241) RGB565 XRGB8888 ARGB8888
interface: 'wl_output',                                  version:  2, name:  4
	x: 0, y: 0, scale: 2,
	physical_width: 370 mm, physical_height: 230 mm,
	make: 'SHP', model: '0x14d6',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 3840 px, height: 2400 px, refresh: 59.993 Hz,
		flags: current preferred
interface: 'zxdg_output_manager_v1',                     version:  3, name:  5
	xdg_output_v1
		output: 4
		name: 'eDP-1'
		description: 'Built-in display'
		logical_x: 0, logical_y: 0
		logical_width: 2560, logical_height: 1600
interface: 'wl_data_device_manager',                     version:  3, name:  6
interface: 'zwp_primary_selection_device_manager_v1',    version:  1, name:  7
interface: 'gtk_primary_selection_device_manager',       version:  1, name:  8
interface: 'wl_subcompositor',                           version:  1, name:  9
interface: 'xdg_wm_base',                                version:  4, name: 10
interface: 'zxdg_shell_v6',                              version:  1, name: 11
interface: 'gtk_shell1',                                 version:  5, name: 12
interface: 'wp_viewporter',                              version:  1, name: 13
interface: 'zwp_pointer_gestures_v1',                    version:  3, name: 14
interface: 'zwp_tablet_manager_v2',                      version:  1, name: 15
interface: 'wl_seat',                                    version:  5, name: 16
	name: seat0
	capabilities: pointer keyboard touch
	keyboard repeat rate: 33
	keyboard repeat delay: 500
interface: 'zwp_relative_pointer_manager_v1',            version:  1, name: 17
interface: 'zwp_pointer_constraints_v1',                 version:  1, name: 18
interface: 'zxdg_exporter_v1',                           version:  1, name: 19
interface: 'zxdg_importer_v1',                           version:  1, name: 20
interface: 'zwp_linux_dmabuf_v1',                        version:  4, name: 21
	formats:
	'AR4H'(0x48345241), modifier: 0x00ffffffffffffff
	'XR4H'(0x48345258), modifier: 0x00ffffffffffffff
	'XB4H'(0x48344258), modifier: 0x00ffffffffffffff
	'AB4H'(0x48344241), modifier: 0x00ffffffffffffff
	'RG16'(0x36314752), modifier: 0x00ffffffffffffff
	'XB30'(0x30334258), modifier: 0x00ffffffffffffff
	'XR30'(0x30335258), modifier: 0x00ffffffffffffff
	'AB30'(0x30334241), modifier: 0x00ffffffffffffff
	'AR30'(0x30335241), modifier: 0x00ffffffffffffff
	'XB24'(0x34324258), modifier: 0x00ffffffffffffff
	'XR24'(0x34325258), modifier: 0x00ffffffffffffff
	'AB24'(0x34324241), modifier: 0x00ffffffffffffff
	'AR24'(0x34325241), modifier: 0x00ffffffffffffff
interface: 'zwp_keyboard_shortcuts_inhibit_manager_v1',  version:  1, name: 22
interface: 'zwp_text_input_manager_v3',                  version:  1, name: 23
interface: 'wp_presentation',                            version:  1, name: 24
	presentation clock id: 1 (CLOCK_MONOTONIC)
interface: 'xdg_activation_v1',                          version:  1, name: 25
`

Wayland-info on sway:

`interface: 'wl_shm',                                     version:  1, name:  1
	formats: 'AB4H'(0x48344241) 'XB4H'(0x48344258) 'AB30'(0x30334241) 'XB30'(0x30334258) RGB565 'RA15'(0x35314152) 'RX15'(0x35315852) 'RA12'(0x32314152) 'RX12'(0x32315852) 'BG24'(0x34324742) 'AB24'(0x34324241) 'XB24'(0x34324258) XRGB8888 ARGB8888
interface: 'wl_drm',                                     version:  2, name:  2
interface: 'zwp_linux_dmabuf_v1',                        version:  4, name:  3
	formats:
	'Y216'(0x36313259), modifier: 0x0100000000000002
	'Y216'(0x36313259), modifier: 0x0100000000000001
	'Y216'(0x36313259), modifier: 0x0000000000000000
	'Y216'(0x36313259), modifier: 0x00ffffffffffffff
	'Y212'(0x32313259), modifier: 0x0100000000000002
	'Y212'(0x32313259), modifier: 0x0100000000000001
	'Y212'(0x32313259), modifier: 0x0000000000000000
	'Y212'(0x32313259), modifier: 0x00ffffffffffffff
	'Y210'(0x30313259), modifier: 0x0100000000000002
	'Y210'(0x30313259), modifier: 0x0100000000000001
	'Y210'(0x30313259), modifier: 0x0000000000000000
	'Y210'(0x30313259), modifier: 0x00ffffffffffffff
	'UYVY'(0x59565955), modifier: 0x0100000000000002
	'UYVY'(0x59565955), modifier: 0x0100000000000001
	'UYVY'(0x59565955), modifier: 0x0000000000000000
	'UYVY'(0x59565955), modifier: 0x00ffffffffffffff
	'YUYV'(0x56595559), modifier: 0x0100000000000002
	'YUYV'(0x56595559), modifier: 0x0100000000000001
	'YUYV'(0x56595559), modifier: 0x0000000000000000
	'YUYV'(0x56595559), modifier: 0x00ffffffffffffff
	'Y416'(0x36313459), modifier: 0x0100000000000002
	'Y416'(0x36313459), modifier: 0x0100000000000001
	'Y416'(0x36313459), modifier: 0x0000000000000000
	'Y416'(0x36313459), modifier: 0x00ffffffffffffff
	'Y412'(0x32313459), modifier: 0x0100000000000002
	'Y412'(0x32313459), modifier: 0x0100000000000001
	'Y412'(0x32313459), modifier: 0x0000000000000000
	'Y412'(0x32313459), modifier: 0x00ffffffffffffff
	'Y410'(0x30313459), modifier: 0x0100000000000002
	'Y410'(0x30313459), modifier: 0x0100000000000001
	'Y410'(0x30313459), modifier: 0x0000000000000000
	'Y410'(0x30313459), modifier: 0x00ffffffffffffff
	'XYUV'(0x56555958), modifier: 0x0100000000000002
	'XYUV'(0x56555958), modifier: 0x0100000000000001
	'XYUV'(0x56555958), modifier: 0x0000000000000000
	'XYUV'(0x56555958), modifier: 0x00ffffffffffffff
	'AYUV'(0x56555941), modifier: 0x0100000000000002
	'AYUV'(0x56555941), modifier: 0x0100000000000001
	'AYUV'(0x56555941), modifier: 0x0000000000000000
	'AYUV'(0x56555941), modifier: 0x00ffffffffffffff
	'NV16'(0x3631564e), modifier: 0x0100000000000002
	'NV16'(0x3631564e), modifier: 0x0100000000000001
	'NV16'(0x3631564e), modifier: 0x0000000000000000
	'NV16'(0x3631564e), modifier: 0x00ffffffffffffff
	'P016'(0x36313050), modifier: 0x0100000000000002
	'P016'(0x36313050), modifier: 0x0100000000000001
	'P016'(0x36313050), modifier: 0x0000000000000000
	'P016'(0x36313050), modifier: 0x00ffffffffffffff
	'P012'(0x32313050), modifier: 0x0100000000000002
	'P012'(0x32313050), modifier: 0x0100000000000001
	'P012'(0x32313050), modifier: 0x0000000000000000
	'P012'(0x32313050), modifier: 0x00ffffffffffffff
	'P010'(0x30313050), modifier: 0x0100000000000002
	'P010'(0x30313050), modifier: 0x0100000000000001
	'P010'(0x30313050), modifier: 0x0000000000000000
	'P010'(0x30313050), modifier: 0x00ffffffffffffff
	'NV12'(0x3231564e), modifier: 0x0100000000000002
	'NV12'(0x3231564e), modifier: 0x0100000000000001
	'NV12'(0x3231564e), modifier: 0x0000000000000000
	'NV12'(0x3231564e), modifier: 0x00ffffffffffffff
	'YV24'(0x34325659), modifier: 0x0100000000000002
	'YV24'(0x34325659), modifier: 0x0100000000000001
	'YV24'(0x34325659), modifier: 0x0000000000000000
	'YV24'(0x34325659), modifier: 0x00ffffffffffffff
	'YV16'(0x36315659), modifier: 0x0100000000000002
	'YV16'(0x36315659), modifier: 0x0100000000000001
	'YV16'(0x36315659), modifier: 0x0000000000000000
	'YV16'(0x36315659), modifier: 0x00ffffffffffffff
	'YV12'(0x32315659), modifier: 0x0100000000000002
	'YV12'(0x32315659), modifier: 0x0100000000000001
	'YV12'(0x32315659), modifier: 0x0000000000000000
	'YV12'(0x32315659), modifier: 0x00ffffffffffffff
	'YV11'(0x31315659), modifier: 0x0100000000000002
	'YV11'(0x31315659), modifier: 0x0100000000000001
	'YV11'(0x31315659), modifier: 0x0000000000000000
	'YV11'(0x31315659), modifier: 0x00ffffffffffffff
	'YVU9'(0x39555659), modifier: 0x0100000000000002
	'YVU9'(0x39555659), modifier: 0x0100000000000001
	'YVU9'(0x39555659), modifier: 0x0000000000000000
	'YVU9'(0x39555659), modifier: 0x00ffffffffffffff
	'YU24'(0x34325559), modifier: 0x0100000000000002
	'YU24'(0x34325559), modifier: 0x0100000000000001
	'YU24'(0x34325559), modifier: 0x0000000000000000
	'YU24'(0x34325559), modifier: 0x00ffffffffffffff
	'YU16'(0x36315559), modifier: 0x0100000000000002
	'YU16'(0x36315559), modifier: 0x0100000000000001
	'YU16'(0x36315559), modifier: 0x0000000000000000
	'YU16'(0x36315559), modifier: 0x00ffffffffffffff
	'YU12'(0x32315559), modifier: 0x0100000000000002
	'YU12'(0x32315559), modifier: 0x0100000000000001
	'YU12'(0x32315559), modifier: 0x0000000000000000
	'YU12'(0x32315559), modifier: 0x00ffffffffffffff
	'YU11'(0x31315559), modifier: 0x0100000000000002
	'YU11'(0x31315559), modifier: 0x0100000000000001
	'YU11'(0x31315559), modifier: 0x0000000000000000
	'YU11'(0x31315559), modifier: 0x00ffffffffffffff
	'YUV9'(0x39565559), modifier: 0x0100000000000002
	'YUV9'(0x39565559), modifier: 0x0100000000000001
	'YUV9'(0x39565559), modifier: 0x0000000000000000
	'YUV9'(0x39565559), modifier: 0x00ffffffffffffff
	'GR32'(0x32335247), modifier: 0x0100000000000004
	'GR32'(0x32335247), modifier: 0x0100000000000002
	'GR32'(0x32335247), modifier: 0x0100000000000001
	'GR32'(0x32335247), modifier: 0x0000000000000000
	'GR32'(0x32335247), modifier: 0x00ffffffffffffff
	'GR88'(0x38385247), modifier: 0x0100000000000002
	'GR88'(0x38385247), modifier: 0x0100000000000001
	'GR88'(0x38385247), modifier: 0x0000000000000000
	'GR88'(0x38385247), modifier: 0x00ffffffffffffff
	'R16 '(0x20363152), modifier: 0x0100000000000002
	'R16 '(0x20363152), modifier: 0x0100000000000001
	'R16 '(0x20363152), modifier: 0x0000000000000000
	'R16 '(0x20363152), modifier: 0x00ffffffffffffff
	'R8  '(0x20203852), modifier: 0x0100000000000002
	'R8  '(0x20203852), modifier: 0x0100000000000001
	'R8  '(0x20203852), modifier: 0x0000000000000000
	'R8  '(0x20203852), modifier: 0x00ffffffffffffff
	'RG16'(0x36314752), modifier: 0x0100000000000002
	'RG16'(0x36314752), modifier: 0x0100000000000001
	'RG16'(0x36314752), modifier: 0x0000000000000000
	'RG16'(0x36314752), modifier: 0x00ffffffffffffff
	'AR15'(0x35315241), modifier: 0x0100000000000002
	'AR15'(0x35315241), modifier: 0x0100000000000001
	'AR15'(0x35315241), modifier: 0x0000000000000000
	'AR15'(0x35315241), modifier: 0x00ffffffffffffff
	'XB24'(0x34324258), modifier: 0x0100000000000004
	'XB24'(0x34324258), modifier: 0x0100000000000002
	'XB24'(0x34324258), modifier: 0x0100000000000001
	'XB24'(0x34324258), modifier: 0x0000000000000000
	'XB24'(0x34324258), modifier: 0x00ffffffffffffff
	'XR24'(0x34325258), modifier: 0x0100000000000004
	'XR24'(0x34325258), modifier: 0x0100000000000002
	'XR24'(0x34325258), modifier: 0x0100000000000001
	'XR24'(0x34325258), modifier: 0x0000000000000000
	'XR24'(0x34325258), modifier: 0x00ffffffffffffff
	'AB24'(0x34324241), modifier: 0x0100000000000004
	'AB24'(0x34324241), modifier: 0x0100000000000002
	'AB24'(0x34324241), modifier: 0x0100000000000001
	'AB24'(0x34324241), modifier: 0x0000000000000000
	'AB24'(0x34324241), modifier: 0x00ffffffffffffff
	'AR24'(0x34325241), modifier: 0x0100000000000004
	'AR24'(0x34325241), modifier: 0x0100000000000002
	'AR24'(0x34325241), modifier: 0x0100000000000001
	'AR24'(0x34325241), modifier: 0x0000000000000000
	'AR24'(0x34325241), modifier: 0x00ffffffffffffff
	'AB30'(0x30334241), modifier: 0x0100000000000002
	'AB30'(0x30334241), modifier: 0x0100000000000001
	'AB30'(0x30334241), modifier: 0x0000000000000000
	'AB30'(0x30334241), modifier: 0x00ffffffffffffff
	'XR30'(0x30335258), modifier: 0x0100000000000002
	'XR30'(0x30335258), modifier: 0x0100000000000001
	'XR30'(0x30335258), modifier: 0x0000000000000000
	'XR30'(0x30335258), modifier: 0x00ffffffffffffff
	'AR30'(0x30335241), modifier: 0x0100000000000002
	'AR30'(0x30335241), modifier: 0x0100000000000001
	'AR30'(0x30335241), modifier: 0x0000000000000000
	'AR30'(0x30335241), modifier: 0x00ffffffffffffff
	'XB4H'(0x48344258), modifier: 0x0100000000000004
	'XB4H'(0x48344258), modifier: 0x0100000000000002
	'XB4H'(0x48344258), modifier: 0x0100000000000001
	'XB4H'(0x48344258), modifier: 0x0000000000000000
	'XB4H'(0x48344258), modifier: 0x00ffffffffffffff
	'AB4H'(0x48344241), modifier: 0x0100000000000004
	'AB4H'(0x48344241), modifier: 0x0100000000000002
	'AB4H'(0x48344241), modifier: 0x0100000000000001
	'AB4H'(0x48344241), modifier: 0x0000000000000000
	'AB4H'(0x48344241), modifier: 0x00ffffffffffffff
interface: 'wl_compositor',                              version:  4, name:  4
interface: 'wl_subcompositor',                           version:  1, name:  5
interface: 'wl_data_device_manager',                     version:  3, name:  6
interface: 'zwlr_gamma_control_manager_v1',              version:  1, name:  7
interface: 'zxdg_output_manager_v1',                     version:  3, name:  8
	xdg_output_v1
		output: 41
		name: 'eDP-1'
		description: 'Sharp Corporation 0x14D6 0x00000000 (eDP-1)'
		logical_x: 0, logical_y: 0
		logical_width: 2880, logical_height: 1800
interface: 'org_kde_kwin_idle',                          version:  1, name:  9
interface: 'zwp_idle_inhibit_manager_v1',                version:  1, name: 10
interface: 'zwlr_layer_shell_v1',                        version:  4, name: 11
interface: 'xdg_wm_base',                                version:  2, name: 12
interface: 'zwp_tablet_manager_v2',                      version:  1, name: 13
interface: 'org_kde_kwin_server_decoration_manager',     version:  1, name: 14
interface: 'zxdg_decoration_manager_v1',                 version:  1, name: 15
interface: 'zwp_relative_pointer_manager_v1',            version:  1, name: 16
interface: 'zwp_pointer_constraints_v1',                 version:  1, name: 17
interface: 'wp_presentation',                            version:  1, name: 18
	presentation clock id: 1 (CLOCK_MONOTONIC)
interface: 'zwlr_output_manager_v1',                     version:  2, name: 19
interface: 'zwlr_output_power_manager_v1',               version:  1, name: 20
interface: 'zwp_input_method_manager_v2',                version:  1, name: 21
interface: 'zwp_text_input_manager_v3',                  version:  1, name: 22
interface: 'zwlr_foreign_toplevel_manager_v1',           version:  3, name: 23
interface: 'wp_drm_lease_device_v1',                     version:  1, name: 24
interface: 'zwlr_export_dmabuf_manager_v1',              version:  1, name: 25
interface: 'zwlr_screencopy_manager_v1',                 version:  3, name: 26
interface: 'zwlr_data_control_manager_v1',               version:  2, name: 27
interface: 'zwp_primary_selection_device_manager_v1',    version:  1, name: 28
interface: 'wp_viewporter',                              version:  1, name: 29
interface: 'zxdg_exporter_v1',                           version:  1, name: 30
interface: 'zxdg_importer_v1',                           version:  1, name: 31
interface: 'zxdg_exporter_v2',                           version:  1, name: 32
interface: 'zxdg_importer_v2',                           version:  1, name: 33
interface: 'xdg_activation_v1',                          version:  1, name: 34
interface: 'zwp_virtual_keyboard_manager_v1',            version:  1, name: 35
interface: 'zwlr_virtual_pointer_manager_v1',            version:  2, name: 36
interface: 'zwlr_input_inhibit_manager_v1',              version:  1, name: 37
interface: 'zwp_keyboard_shortcuts_inhibit_manager_v1',  version:  1, name: 38
interface: 'wl_seat',                                    version:  7, name: 39
	name: seat0
	capabilities: pointer keyboard touch
	keyboard repeat rate: 25
	keyboard repeat delay: 600
interface: 'zwp_pointer_gestures_v1',                    version:  3, name: 40
interface: 'wl_output',                                  version:  4, name: 41
	x: 0, y: 0, scale: 1,
	physical_width: 370 mm, physical_height: 230 mm,
	make: 'Sharp Corporation', model: '0x14D6',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 2880 px, height: 1800 px, refresh: 59.969 Hz,
		flags: current
`

**Comment by unrelentingtech** [comment]
So gnome doesn't support modifiers (probably still [this](https://gitlab.gnome.org/GNOME/mutter/-/issues/1229)). This is odd, mesa shouldn't try to create resources with modifiers then.

Can you use a debugger to break on that `fprintf` in `iris_resource_create_with_modifiers` and read the backtrace?

**Comment by szbergeron** [comment]
Would that require rebuilding mesa? I'm also a bit lost on how this would work, I thought the code above ran in a separate process/somehow isolated from the application

**Comment by szbergeron** [comment]
Running in gdb I get an entirely separate issue:

```
(gdb) run
Starting program: /home/sawyer/oss/wezterm/target/debug/wezterm 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7ffff77ce640 (LWP 254912)]
[New Thread 0x7ffff75cd640 (LWP 254913)]
[New Thread 0x7ffff73cc640 (LWP 254914)]
[New Thread 0x7ffff71cb640 (LWP 254915)]
[Thread 0x7ffff71cb640 (LWP 254915) exited]
[Thread 0x7ffff73cc640 (LWP 254914) exited]
[Thread 0x7ffff75cd640 (LWP 254913) exited]
[Thread 0x7ffff77ce640 (LWP 254912) exited]
process 254877 is executing new program: /home/sawyer/oss/wezterm/target/debug/wezterm-gui
warning: Missing auto-load script at offset 0 in section .debug_gdb_scripts
of file /home/sawyer/oss/wezterm/target/debug/wezterm-gui.
Use `info auto-load python-scripts [REGEXP]' to list them.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7ffff71db640 (LWP 255010)]
[New Thread 0x7ffff6fda640 (LWP 255011)]
[New Thread 0x7ffff6dd9640 (LWP 255034)]
[New Thread 0x7ffff6bd8640 (LWP 255035)]
[New Thread 0x7ffff69d7640 (LWP 255036)]
 2022-03-19T20:23:29.452Z INFO  wezterm_mux_server_impl::local > setting up /run/user/1000/wezterm/gui-sock-254877
[New Thread 0x7ffff67d6640 (LWP 255040)]
[Detaching after fork from child process 255042]
[New Thread 0x7ffff65d5640 (LWP 255043)]
[New Thread 0x7ffff63d4640 (LWP 255044)]
[New Thread 0x7ffff61d3640 (LWP 255046)]
[New Thread 0x7ffff5ed1640 (LWP 255048)]
[New Thread 0x7ffff5891640 (LWP 255061)]
[New Thread 0x7ffff5690640 (LWP 255062)]
[New Thread 0x7ffff548c640 (LWP 255063)]
[New Thread 0x7ffff5288640 (LWP 255064)]
[New Thread 0x7ffff5084640 (LWP 255065)]
[New Thread 0x7ffff4e80640 (LWP 255066)]
[New Thread 0x7ffff4c7c640 (LWP 255067)]
[New Thread 0x7ffff4a78640 (LWP 255069)]
[New Thread 0x7ffff4874640 (LWP 255070)]
[New Thread 0x7ffff4673640 (LWP 255071)]
[New Thread 0x7ffff446c640 (LWP 255072)]
[New Thread 0x7ffff4268640 (LWP 255073)]
[Detaching after fork from child process 255256]
[New Thread 0x7fff9c89c640 (LWP 255287)]
[New Thread 0x7fff8e042640 (LWP 255288)]
[New Thread 0x7fff8d841640 (LWP 255289)]
[New Thread 0x7fff8d040640 (LWP 255290)]
[New Thread 0x7fff8c83f640 (LWP 255291)]
[New Thread 0x7fff7ffff640 (LWP 255292)]
[New Thread 0x7fff7f7fe640 (LWP 255293)]
[New Thread 0x7fff7effd640 (LWP 255294)]
[New Thread 0x7fff7e7fc640 (LWP 255295)]
[New Thread 0x7fff7dffb640 (LWP 255296)]
[New Thread 0x7fff7d7fa640 (LWP 255297)]
Unsupported modifier, resource creation failed.
XXX: resource creation failed
 2022-03-19T20:23:36.011Z INFO  wezterm_gui::termwindow        > OpenGL initialized! Mesa Intel(R) UHD Graphics (CML GT2) 4.6 (Compatibility Profile) Mesa 22.0.0 is_context_loss_possible=false wezterm version: 20220317-165235-27088fab
[New Thread 0x7fff7c979640 (LWP 255298)]
wl_surface@17: error 2: Buffer size (820x593) must be an integer multiple of the buffer_scale (2)
 2022-03-19T20:23:36.173Z ERROR wezterm_gui                    > error during event_q.dispatch protocol_error=Some(ProtocolError { code: 2, object_id: 17, object_interface: "wl_surface", message: "" }): Protocol error (os error 71); terminating
[New Thread 0x7fff7c4b8640 (LWP 255299)]
thread 'main' panicked at 'cannot access a Thread Local Storage value during or after destruction: AccessError', /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/local.rs:388:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[New Thread 0x7fff7c2b7640 (LWP 255300)]
[Thread 0x7fff7d7fa640 (LWP 255297) exited]
[Thread 0x7fff7dffb640 (LWP 255296) exited]
[Thread 0x7fff7e7fc640 (LWP 255295) exited]
[Thread 0x7fff7effd640 (LWP 255294) exited]
[Thread 0x7fff7f7fe640 (LWP 255293) exited]
[Thread 0x7fff7ffff640 (LWP 255292) exited]
[Thread 0x7fff8c83f640 (LWP 255291) exited]
[Thread 0x7fff8d040640 (LWP 255290) exited]
[Thread 0x7fff8d841640 (LWP 255289) exited]
[Thread 0x7fff8e042640 (LWP 255288) exited]
[Thread 0x7fff9c89c640 (LWP 255287) exited]

Thread 1 "wezterm-gui" received signal SIGILL, Illegal instruction.
std::thread::local::fast::destroy_value<core::cell::RefCell<core::option::Option<alloc::rc::Rc<wezterm_gui::frontend::GuiFrontEnd>>>> (ptr=0x7ffff71dcc00) at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/local.rs:652
652	/rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/local.rs: No such file or directory.
(gdb)
```

**Comment by szbergeron** [comment]
Further experimentation: if I don't set the breakpoint I get the failure from earlier. If I do set the breakpoint I see the issue directly above

**Comment by unrelentingtech** [comment]
The "Buffer size (820x593) must be an integer multiple of the buffer_scale (2)"? Weird, but that's all after "Unsupported modifier, resource creation failed" anyway.

Yeah, mesa with debug symbols is needed, there are [ways to get them for distro packages in some linux distros](https://wiki.archlinux.org/title/Debuginfod) apparently but I would just install a fully debug built mesa from git.

> I thought the code above ran in a separate process/somehow isolated from the application

No, how would it be isolated? Any IPC in a graphics API would make everything slow. It's all one call stack from the highest level GL/Vulkan calls into the driver that uses ioctls to talk to the kernel driver.

**Comment by szbergeron** [comment]
It seems this pulls in mesa dynamically, I think I've built and installed a mesa debug build but trying to set a breakpoint there in gdb doesn't actually do anything.

I built mesa from the aur mesa-git package with `-D buildtype=debug` substituted in to the meson setup command

**Comment by unrelentingtech** [comment]
Yes, typically on modern unix systems libGL is actually [glvnd](https://gitlab.freedesktop.org/glvnd/libglvnd) which dynamically loads mesa (or nvidia). And applications commonly load libGL itself via dlopen too.

I think gdb should support setting a breakpoint on a symbol name that is not loaded yet (at least without the line number)? But in any case you can first break on something in between loading GL and the error. (somewhere [here](https://github.com/wez/wezterm/blob/aba3da6a02141760dba7c2b42707d7ef7a41b30b/window/src/os/wayland/window.rs#L705) maybe)

**Comment by szbergeron** [comment]
Managed to break my system a couple times trying to debug this so I think this one may be a bit over my head at the moment (never did get gdb to break anywhere inside mesa, haven't had luck getting gdb to recognize wezterm source files in a way that lets me break on the line in question)

**Comment by szbergeron** [comment]
This seems to not happen if I have fractional scaling disabled, btw

**Comment by 1player** [comment]
Sorry for double posting, my issue looks a little different. It doesn't crash with SIGILL, but SIGABRT in my case:

```
09:44:08.309  INFO   wezterm_mux_server_impl::local > setting up /run/user/1000/wezterm/gui-sock-107442
09:44:08.360  INFO   wezterm_gui::termwindow        > OpenGL initialized! AMD Radeon RX 6800 XT (sienna_cichlid, LLVM 13.0.1, DRM 3.44, 5.17.3-300.fc36.x86_64) 4.6 (Compatibility Profile) Mesa 22.0.1 is_context_loss_possible=false wezterm version: 20220408-101518-b908e2dd
09:44:08.413  ERROR  wezterm_gui                    > Error while flushing display: Broken pipe (os error 32); terminating
thread 'main' panicked at 'cannot access a Thread Local Storage value during or after destruction: AccessError', /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/thread/local.rs:389:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'panic in a function that cannot unwind', /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/thread/local.rs:653:5
stack backtrace:
   0:     0x5561e22dc4dd - <unknown>
   1:     0x5561e2307c9c - <unknown>
   2:     0x5561e22d2738 - <unknown>
   3:     0x5561e22dee07 - <unknown>
   4:     0x5561e22dead0 - <unknown>
   5:     0x5561e22df594 - <unknown>
   6:     0x5561e22df209 - <unknown>
   7:     0x5561e22dc9a4 - <unknown>
   8:     0x5561e22def59 - <unknown>
   9:     0x5561e0f685eb - <unknown>
  10:     0x5561e11486c4 - <unknown>
  11:     0x7febc197b82f - __GI___call_tls_dtors
  12:     0x7febc197b1e9 - __run_exit_handlers
  13:     0x7febc197b230 - exit
  14:     0x5561e22e38d7 - <unknown>
  15:     0x5561e22da5ef - <unknown>
  16:     0x5561e10e01d7 - <unknown>
  17:     0x5561e10e0265 - <unknown>
  18:     0x5561e10e0339 - <unknown>
  19:     0x5561e11345b3 - <unknown>
  20:     0x5561e103b249 - <unknown>
  21:     0x5561e22dbbc1 - <unknown>
  22:     0x5561e10ffbf2 - <unknown>
  23:     0x7febc1963590 - __libc_start_call_main
  24:     0x7febc1963649 - __libc_start_main_alias_2
  25:     0x5561e0f68eb5 - <unknown>
  26:                0x0 - <unknown>
thread panicked while panicking. aborting.
fish: Job 1, 'wezterm' terminated by signal SIGABRT (Abort)
```

**Comment by 1player** [comment]
Can't replicate on latest nightly.

**Comment by sandangel** [comment]
I think the issue happen when you scale your display to 200% on wayland. when I change it to 100%, the terminal can launch
my workaround for this is I use display 100%, then set the font size to really big (20).

**Comment by 1player** [comment]
@sandangel have you tried the latest nightly? Because I'm on 200% scaling and Wayland enabled, and it's not crashing as per my previous comment.

**Comment by sandangel** [comment]
@1player I have tried latest nightly. It depends on your monitor though. This is my error:

```sh
wl_surface@17: error 2: Buffer size (1458x1175) must be an integer multiple of the buffer_scale (2)
```
when I have 200% scale.

**Comment by sandangel** [comment]
as I understand 1175 is not an integer multiple of buffer_scale, hence it crashed.

**Comment by sandangel** [comment]
my resolution is 4112x2572 (M1 mac, with VM running fedora)

**Comment by wez** [comment]
re: the `must be an integer multiple of the buffer_scale` error message, I've pushed a commit that resolves that for me.
That wasn't the original issue reported here though; I have no idea about that one!

**Comment by sandangel** [comment]
cool. I tried that commit and it works now with `enable_wayland = true`. thanks a lottttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt

**Comment by sandangel** [comment]
@wez there is a same error message in the issue author debug log comment: https://github.com/wez/wezterm/issues/1727#issuecomment-1073075761

**Comment by wez** [comment]
@szbergeron: would you mind trying the latest nightly to see if things are working better?

**Comment by unrelentingtech** [comment]
Looks like Firefox [had a very similar issue recently](https://bugzilla.mozilla.org/show_bug.cgi?id=1767916). Also only on gnome. ..and yeah they did pretty much what f72207e7f07 does: attach a NULL buffer before changing scale factor.

**Comment by unrelentingtech** [comment]
That commit, however, broke everything on wlroots :/

**Comment by unrelentingtech** [comment]
Arghh. So in wlroots a commit with a null buffer counts as an unmap, which sets the state to unconfigured. Seems like this corresponds to the following bit of the spec:

>  For an xdg_surface to be mapped by the compositor, the following conditions must be met: (1) the client has assigned an xdg_surface-based role to the surface (2) the client has set and committed the xdg_surface state and the role-dependent state to the surface (3) the client has committed a buffer to the surface
>
> A newly-unmapped surface is considered to have met condition (1) out of the 3 required conditions for mapping a surface if its role surface has not been destroyed, i.e. the client must perform the initial commit again before attaching a buffer.

so after this null buffer is attached we need to actually wait for a configure before trying to paint.

**Comment by wez** [comment]
I think the wlroots part of this is fixed in 8f222d9559820b2e63f72a3a9a492a2be2d420ef

**Comment by szbergeron** [comment]
I can't currently reproduce this on my end so this may be fixed. Thanks!
